### PR TITLE
Bootstarp

### PR DESCRIPTION
**Pull request by hira-iqbal**

### CODE REVIEW

File: app/controllers/articles_controller.rb [hunk 1]
```diff
@@ -19,7 +19,7 @@ def edit; end
   def create
     @article = Article.new(article_params)
     if @article.save
-      redirect_to @article
+      redirect_to @article, notice: "article is saved"
      else
       render :new
```
Comment (aliammaar) [inline app/controllers/articles_controller.rb]: render with `object.errors` here.

File: app/controllers/articles_controller.rb [hunk 2]
```diff
@@ -29,7 +29,7 @@ def update
     if @article.update(article_params)
       redirect_to @article
     else
-      render edit
+      render :edit
```
Comment (aliammaar) [inline app/controllers/articles_controller.rb]: same as above^

File: app/controllers/roles_controller.rb [hunk 3]
```diff
@@ -3,9 +3,9 @@ class RolesController < ApplicationController
   def create
     @role = Role.new(role_params)
     if @role.save
-      redirect_to @role
+      redirect_to @role, notice: "Role is saved successfully."
     else
-      render new
+      render :new
```
Comment (aliammaar) [inline app/controllers/roles_controller.rb]: same as above^

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,29 +1,48 @@
...
 
-  def edit
+  def show; end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      redirect_to @user, notice: "User successfully created."
+     else
+      render :new, alert: @user.errors.full_messages
```
Comment (aliammaar) [inline app/controllers/users_controller.rb]: 👍

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,29 +1,48 @@
...
 
   def update
+    if @user.update(user_params)
+       redirect_to @user
+     else
+      render :edit, alert: @user.errors.full_messages
+    end
+  end
+  def destroy
+    @user.destroy
```
Comment (aliammaar) [inline app/controllers/users_controller.rb]: please handle success and failure cases

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,29 +1,48 @@
...
-      @user.update(user_params)
-    end
+    return redirect_to root_path, alert: "this not a valid user" if @user.nil?
   end
 
- def user_params
-    params.require(:user).permit(:name, :surname, :password, :role_id)
+  def user_params
+    params.require(:user).permit(:name, :surname, :email, :password, :role_id)
  end
```
Comment (aliammaar) [inline app/controllers/users_controller.rb]: fix indentation!

File: app/views/articles/_form.html.erb [hunk 7]
```diff
@@ -1,27 +1,26 @@
...
+    <% end %>
+  </div><br>
 
-  <div>
+  <div class="form-group col-md-4">
     <%= form.label :status %>
-    <%= form.select :status, ['public', 'private', 'archived'], selected: 'public' %>
-
-  </div>
+    <%= form.select(:status, Article::STATUSES.map {|e|[e.camelize, e.downcase]},{include_blank: "please select a status"}, {class: "form-control"} ) %>
```
Comment (aliammaar) [inline app/views/articles/_form.html.erb]: move this to a helper method.

File: db/migrate/20211129112149_add_role_to_user.rb [hunk 8]
```diff
@@ -1,5 +1,5 @@
 class AddRoleToUser < ActiveRecord::Migration[6.1]
   def change
-    add_column :users, :role_id, :integer
+    add_references :users, :role_id, :integer
```
Comment (aliammaar) [inline db/migrate/20211129112149_add_role_to_user.rb]: this should bring changes to `db/schema.rb` file. Let's try `rails db:rollback && rails db:migrate` and check in any changes.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (hira-iqbal) [reply app/controllers/users_controller.rb]: I do not understand it.

File: app/controllers/articles_controller.rb (hunk 1 above)
Comment (hira-iqbal) [reply app/controllers/articles_controller.rb]: updated

File: app/controllers/articles_controller.rb (hunk 2 above)
Comment (hira-iqbal) [reply app/controllers/articles_controller.rb]: updated

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (hira-iqbal) [reply app/controllers/users_controller.rb]: Indentation has been fixed.

File: app/views/articles/_form.html.erb (hunk 7 above)
Comment (hira-iqbal) [reply app/views/articles/_form.html.erb]: How do I move it to a helper method?

Comment (hira-iqbal) [reply app/views/articles/_form.html.erb]: Moved !

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -1,2 +1,9 @@
 module ApplicationHelper
+
+  def status_options
+    Article::STATUSES.map {|e|[e.camelize, e.downcase]}
+  end
 end
+
```
Comment (hira-iqbal) [inline app/helpers/application_helper.rb]: need to remove these lines.

File: app/controllers/articles_controller.rb [hunk 10]
```diff
@@ -1,8 +1,8 @@
 class ArticlesController < ApplicationController
-
- before_action :set_article, only: [:show, :edit, :update, :destroy]
- http_basic_authenticate_with name: "dhh", password: "secret", except: [:index, :show]
-
+  before_action :authenticate_user!
+  before_action :set_article, only: [:show, :edit, :update, :destroy]
+  http_basic_authenticate_with name: "dhh", password: "secret", except: [:index, :show]
+  after_action :verify_authorized, except: :index
```
Comment (hira-iqbal) [inline app/controllers/articles_controller.rb:4]: need to replace with `before_action`

File: app/controllers/comments_controller.rb [hunk 11]
```diff
@@ -1,15 +1,17 @@
 class CommentsController < ApplicationController
   http_basic_authenticate_with name: "adha", password: "security", only: [:destroy]
+  before_action :find_article
+  after_action :verify_authorized
```
Comment (hira-iqbal) [inline app/controllers/comments_controller.rb:4]: need to replace with `before_action`

File: app/controllers/articles_controller.rb (hunk 10 above)
Comment (hira-iqbal) [reply app/controllers/articles_controller.rb:4]: updated

File: app/controllers/comments_controller.rb (hunk 11 above)
Comment (hira-iqbal) [reply app/controllers/comments_controller.rb:4]: replaced

File: app/controllers/articles_controller.rb [hunk 12]
```diff
@@ -1,17 +1,18 @@
 class ArticlesController < ApplicationController
-
- before_action :set_article, only: [:show, :edit, :update, :destroy]
- http_basic_authenticate_with name: "dhh", password: "secret", except: [:index, :show]
-
+  before_action :authenticate_user!
+  before_action :set_article, only: [:show, :edit, :update, :destroy]
+  http_basic_authenticate_with name: "dhh", password: "secret", except: [:index, :show]
```
Comment (aliammaar) [inline app/controllers/articles_controller.rb]: I think user is now supposed to login with password. So, we don’t need this.

Comment (hira-iqbal) [reply app/controllers/articles_controller.rb]: Updated

**Pull request merged**